### PR TITLE
HDDS-9898. Recon SCM doesn't store EC containers.

### PR DESCRIPTION
**Pull request by ArafatKhan2198**
## What changes were proposed in this pull request?

Recon utilises a background task known as `PipeLineSyncTask` to manage tasks like creating and removing pipelines. This task operates discreetly in the background, executing every 5 minutes. During Recon's initial startup, the task is triggered, successfully registering existing Ratis Pipelines.

However, when utilising EC replication to create a key, a separate EC pipeline is generated. While this EC pipeline quickly establishes itself in the cluster, Recon initially avoid's registering it. The registration takes place during the periodic execution of the PipelineSyncTask every 5 minutes.

To address the delay in pipeline registration, we've made modifications to the code within `ReconContainerManager`. Now, when adding a new container to Recon, the code checks if the corresponding pipeline is present. If not, instead of bypassing the addition, the code creates the missing pipeline. This adjustment ensures that Recon promptly registers the new EC pipeline, reducing the delay.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-9898
## How was this patch tested?

I have created unit tests for it and have also manually tested it using the local Docker setup.

### CODE REVIEW

**Comment by ArafatKhan2198** [comment]
@dombizita @devmadhuu please have a look.

**Review by sumitagrawl — APPROVED** [review]
@ArafatKhan2198 LGTM

**Comment by kerneltime** [comment]
cc @tanvipenumudy

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconContainerManager.java [hunk 1]
```diff
@@ -258,6 +262,26 @@ public void addNewContainer(ContainerWithPipeline containerWithPipeline)
     }
   }
 
+  /**
+   * Add a new pipeline to the pipeline metadata.
+   *
+   * @param pipeline pipeline
+   * @throws IOException
+   */
+  public void addPipeline(Pipeline pipeline)
```
Comment (devmadhuu) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconContainerManager.java:271]: @ArafatKhan2198 thanks for working on this patch. You don't need to add this new method here, You can directly call `addPipeline` method on existing `pipeLineManager` instance above.

Comment (ArafatKhan2198) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconContainerManager.java]: Thanks done!

**Review by devmadhuu — APPROVED** [review]
Thanks @ArafatKhan2198 for addressing comments. LGTM +1

**Comment by dombizita** [comment]
Thanks for working on this @ArafatKhan2198! Your comments on the jira are really useful for the context, thanks for the detailed analysis there!!

Please correct me if I'm wrong, but in the end the current functionality is correct, right? Right now the Ratis pipelines are pre-created, so during Recon's startup when the pipeline sync task runs it will register those pipelines. So if Recon sees a Ratis container associated with the initially created pipelines everything will be fine, the container will be added to Recon. But I believe if later we have new Ratis pipelines this problem will also occur, like with EC.
The problem is that with EC pipelines this will always happen, as default we don't have any, we create them only when we want to create a key with EC (right?). But after 5 minutes (or the value of the config) when the pipeline sync task ran the EC pipeline will be there in Recon and the container will be added with the pipeline. 
I think the title of the jira is misleading, as SCM stores the EC containers, only problem is that it is delayed because of the pipeline sync task. What do you think of this @ArafatKhan2198 @xBis7, is this correct? 

I tried to look into the history of this code you changed and why was this part there like that before:
https://github.com/apache/ozone/blob/9f475809d24598cd2b99fa62af93e638ab31860b/hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconContainerManager.java#L240-L245
But I couldn't find the reason. My concern is that maybe there are some cases we can't think of right now that will also be affected with this change and we will add pipelines to Recon outside of the pipeline sync task. What do you think, this won't cause problems?

**Comment by xBis7** [comment]
I hadn't followed up on this. 

From @ArafatKhan2198's comments and what I see on master, the actual issue is that every time a new pipeline is created, we have to wait for the `PipelineSyncTask` to run, so that the pipeline is registered. After that, the containers associated with that pipeline, will no longer be considered unknown and will be stored.

> But I believe if later we have new Ratis pipelines this problem will also occur, like with EC.

@dombizita You are right, this will occur for every new pipeline creation. When I created the jira, I hadn't done an analysis to see that Ratis pipelines are pre-created and that makes Recon pickup the containers associated with them immediately.

Based on all that, I think we need to reconsider the approach on this PR. Since Recon stores pipelines using the sync task, doing it in a second place in the code, creates confusion and it can also lead to bugs.

**Comment by adoroszlai** [comment]
Please check unit test failure:

```
Tests run: 10, Failures: 1, Errors: 1, Skipped: 0, Time elapsed: 16.12 s <<< FAILURE! -- in org.apache.hadoop.ozone.recon.api.TestEndpoints
org.apache.hadoop.ozone.recon.api.TestEndpoints.testGetPipelines -- Time elapsed: 5.809 s <<< ERROR!
NullPointerException
	at org.apache.hadoop.ozone.recon.api.PipelineEndpoint.getLastLeaderElectionElapsedTimeMetricValue(PipelineEndpoint.java:163)
	at org.apache.hadoop.ozone.recon.api.PipelineEndpoint.lambda$getPipelines$1(PipelineEndpoint.java:120)
	at java.util.ArrayList.forEach(ArrayList.java:1259)
	at org.apache.hadoop.ozone.recon.api.PipelineEndpoint.getPipelines(PipelineEndpoint.java:77)
	at org.apache.hadoop.ozone.recon.api.TestEndpoints.testGetPipelines(TestEndpoints.java:635)

org.apache.hadoop.ozone.recon.api.TestEndpoints.testGetDatanodes -- Time elapsed: 0.887 s <<< FAILURE!
AssertionFailedError: expected: <1> but was: <0>
	at org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)
	at org.junit.jupiter.api.AssertionFailureBuilder.buildAndThrow(AssertionFailureBuilder.java:132)
	at org.junit.jupiter.api.AssertEquals.failNotEqual(AssertEquals.java:197)
	at org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:150)
	at org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:145)
	at org.junit.jupiter.api.Assertions.assertEquals(Assertions.java:531)
	at org.apache.hadoop.ozone.recon.api.TestEndpoints.testDatanodeResponse(TestEndpoints.java:543)
	at org.apache.hadoop.ozone.recon.api.TestEndpoints.lambda$testGetDatanodes$0(TestEndpoints.java:574)
	at java.util.ArrayList.forEach(ArrayList.java:1259)
	at org.apache.hadoop.ozone.recon.api.TestEndpoints.testGetDatanodes(TestEndpoints.java:572)
```

**Comment by sumitagrawl** [comment]
> I hadn't followed up on this.
> 
> From @ArafatKhan2198's comments and what I see on master, the actual issue is that every time a new pipeline is created, we have to wait for the `PipelineSyncTask` to run, so that the pipeline is registered. After that, the containers associated with that pipeline, will no longer be considered unknown and will be stored.
> 
> > But I believe if later we have new Ratis pipelines this problem will also occur, like with EC.
> 
> @dombizita You are right, this will occur for every new pipeline creation. When I created the jira, I hadn't done an analysis to see that Ratis pipelines are pre-created and that makes Recon pickup the containers associated with them immediately.
> 
> Based on all that, I think we need to reconsider the approach on this PR. Since Recon stores pipelines using the sync task, doing it in a second place in the code, creates confusion and it can also lead to bugs.

Recon is evolved from SCM having the strict check. Can not find any background but have fixed multiple issue for adding container like container in closed state.

Can not find any impact with this logic, as here, its getting pipeline information from SCM associated with container, same is used to update the pipeline and then container.

**Comment by sumitagrawl** [comment]
> There is no impact and I'm not saying that we shouldn't proceed with the changes. But there could be better solutions that also keep consistency. For example, we can trigger a pipeline sync before adding new containers with batch.

@xBis7 Related to above points,
SCM do not have this problem as pipeline is created before container creation.

Pipeline is added to recon in below 2 cases, both confirm with SCM before adding pipeline to Recon:
1. Pipeline sync task triggered at 5 min to ensure scm and recon pipelines are sync
2. ReconPipelineReportHandler: DN trigger Pipeline Report every 1 min, and pipeline is added to recon using this
But container add is triggered through ICR by DN for every container to Recon.

For Ratis, pipeline is created and marked open in Pipeline Report from DN, so Recon do not face issue for ratis as SCM also wait for same event.
For EC, problem is observed as SCM directly mark as OPEN and no way to sync this to recon.

Now, ICR for container add can be more frequent and for every ICR, triggering pipeline task sync will not be appropriate. This will keep buffering the ICR at Recon for large cluster.

Now for container add, it always sync with SCM to get container with pipeline, and to avoid buffering, its optimised to add the pipeline details obtained with container itself. This is safe and consistent as pipeline is obtained from SCM itself.

**Comment by xBis7** [comment]
@sumitagrawl Thanks for the explanation. Based on all that, I'm fine with this PR.

**Pull request merged**